course5/homework12/main.cc

```diff
@@ -0,0 +1,164 @@
+        m_descending{descending}
+        , m_column{column}
+        , m_separator{separator}
+        {}
+
+    bool operator() (const std::string& s1, const std::string& s2)
+    {
+        std::string to_cmp_1 = s1;
+        std::string to_cmp_2 = s2;
+
+        std::vector<std::string> split_s1;
+        std::vector<std::string> split_s2;
+
+        if (m_column != 0 && m_separator == "")
+            m_separator = " ";
+        if (m_column != 0)
+        {
+            boost::split(split_s1, s1, boost::is_any_of(m_separator));
+            boost::split(split_s2, s2, boost::is_any_of(m_separator));
+
+            to_cmp_1 = split_s1[m_column];
+            to_cmp_2 = split_s2[m_column];
+        }
+
+        if (m_descending)
+            return to_cmp_1 > to_cmp_2;
+        return to_cmp_1 < to_cmp_2;
+    }
+
+    bool m_descending = false;
+    int m_column = 0;
+    std::string m_separator = "";
+};
+
+
+int main(int argc, char** argv)
+{
+    namespace po = boost::program_options;
+
+    bool descending = false;
+    std::string separator = "";
+    int column = 0;
+
+    po::options_description desc("Sort input according to parameters.\nAllowed options:");
+    desc.add_options()
+        ("help,h", "Show this message")
+        ("input,i", po::value<std::string>(), "input file (defaults to stdin)")
+        ("output,o", po::value<std::string>(), "output file (defaults to stdout)")
+        ("desc,d", po::bool_switch(&descending), "sort input in descending order (defaults to ascending)")
+        ("col,c", po::value<int>(), "sort input lines by <col>th column (counting from 0), as split by <sep>")
+        ("sep,s", po::value<std::string>(), "break input lines on sep characters (defaults to <space>, if col is specified)");
+        // (help, file, asc/desc, field sep, col num, numeric)
+    ;
+
+    po::positional_options_description p;
+    p.add("input", -1);
+
+    po::variables_map vm;
+    po::store(po::command_line_parser(argc, argv).options(desc).positional(p).run(), vm);
+    po::notify(vm);
+
+    if (vm.count("help")) {
+        std::cout << desc << std::endl;;
+        return 1;
+    }
+
+    if (vm.count("col")) {
+        column = vm["col"].as<int>();
+    }
+
+    if (vm.count("sep")) {
+        separator = vm["sep"].as<std::string>();
+    }
+    
+    std::unique_ptr<std::vector<std::string>> buffer = std::make_unique<std::vector<std::string>>();
+    if (vm.count("input"))
+    {
+        std::string in_fname = vm["input"].as<std::string>();
+        std::ifstream infile(in_fname);
+        if (infile)
+        {
+            read(infile, *buffer.get());
+        }
+        else
+        {
+            std::string msg = std::strerror(errno);
+            throw std::runtime_error("Failed to open file: " + msg);
+        }
+    }
+    else
+    {
+        read(std::cin, *buffer.get());
+    }
+
+    boost::sort(*buffer.get(), sorter(descending, column, separator));
```

Comment:
```unique_ptr``` has all the needed operators overloaded. In your code you use it just like if it was a raw pointer.
So in this case: ```*buffer```